File: RewardSampleForObjC/RewardSampleForObjC/VideoSingleViewController.m
//
//  VideoSingleViewController.m
//  RewardSampleForObjC
//
//  Created by AdGeneration on 2017/02/21.
//  Copyright © 2017年 Supership Inc. All rights reserved.
//

#import <AVFoundation/AVFoundation.h>
#import <VAMP/VAMP.h>

#import "VideoSingleViewController.h"

@interface VideoSingleViewController () <VAMPDelegate>

@property (nonatomic, weak) IBOutlet UITextField *adcodeField;
@property (nonatomic, weak) IBOutlet UITextView *adLogView;

@property (nonatomic) UIBarButtonItem *soundOffButton;
@property (nonatomic) UIBarButtonItem *soundOnButton;
@property (nonatomic) AVAudioPlayer *soundPlayer;
@property (nonatomic) BOOL isPlayingPrev;

@property (nonatomic, readonly) NSString *placementId;
@property (nonatomic) VAMP *vamp;

@end

@implementation VideoSingleViewController

/**
 広告枠IDを設定してください
   59755 : iOSテスト用ID (このIDのままリリースしないでください)
 */
static NSString * const kPlacementId = @"59755";

- (void)viewDidLoad {
    [super viewDidLoad];
    
    // テストモードか確認します
    NSLog(@"[VAMP]isTestMode:%@", [VAMP isTestMode] ? @"YES" : @"NO");
    
    // デバッグモードか確認します
    NSLog(@"[VAMP]isDebugMode:%@", [VAMP isDebugMode] ? @"YES" : @"NO");
    
    self.automaticallyAdjustsScrollViewInsets = NO;
    
    self.adcodeField.text = self.placementId;
    
    self.adLogView.text = @"";
    self.adLogView.editable = NO;
    
    UIImage *soundOnImage = [[UIImage imageNamed:@"soundOn"] imageWithRenderingMode:UIImageRenderingModeAlwaysOriginal];
    self.soundOffButton = [[UIBarButtonItem alloc] initWithImage:soundOnImage
                                                           style:UIBarButtonItemStylePlain
                                                          target:self
                                                          action:@selector(soundOff)];
    
    UIImage *soundOffImage = [[UIImage imageNamed:@"soundOff"] imageWithRenderingMode:UIImageRenderingModeAlwaysOriginal];
    self.soundOnButton = [[UIBarButtonItem alloc] initWithImage:soundOffImage
                                                          style:UIBarButtonItemStylePlain
                                                         target:self
                                                         action:@selector(soundOn)];
    
    self.navigationItem.rightBarButtonItem = self.soundOnButton;
    
    NSString *soundPath = [[NSBundle mainBundle] pathForResource:@"invisible" ofType:@"mp3"];
    NSURL *soundUrl = [NSURL fileURLWithPath:soundPath];
    
    NSError *error = nil;
    self.soundPlayer = [[AVAudioPlayer alloc] initWithContentsOfURL:soundUrl error:&error];
    [self.soundPlayer prepareToPlay];
    
    if (error) {
        NSLog(@"sound player error:%@", error);
    }
    
    // VAMPインスタンスを生成し初期化します
    self.vamp = [VAMP new];
    self.vamp.delegate = self;
    [self.vamp setPlacementId:self.placementId];
    [self.vamp setRootViewController:self];
}

- (NSString *)placementId {
    return kPlacementId;
}

- (void)soundOff {
    self.navigationItem.rightBarButtonItem = self.soundOnButton;
    [self.soundPlayer pause];
}

- (void)soundOn {
    self.navigationItem.rightBarButtonItem = self.soundOffButton;
    [self.soundPlayer play];
}

- (IBAction)loadAd:(id)sender {
    // 広告の読み込みを開始します
    [self.vamp load];
    
    [self addLogText:@"[load]"];
}

- (IBAction)showAd:(id)sender {
    // 広告の準備ができているか確認してから表示してください
    if (self.vamp.isReady) {
        // 広告を再生します
        BOOL success = [self.vamp show];
        
        [self addLogText:@"[show]"];
        
        self.isPlayingPrev = self.soundPlayer.isPlaying;
        
        if (self.soundPlayer.isPlaying && success) {
            [self.soundPlayer pause];
        }
    } else {
        NSLog(@"[VAMP]not ready");
    }
}

- (IBAction)clearLoadedAd:(id)sender {
    // ロード済みの広告を破棄します。このメソッドを実行した後はloadからやり直してください
    [self.vamp clearLoaded];
    
    [self addLogText:@"[clear]"];
}

- (void)addLogText:(NSString *)message {
    NSDate *now = [NSDate date];
    NSDateFormatter *dateFormatter = [[NSDateFormatter alloc] init];
    dateFormatter.locale = [NSLocale systemLocale];
    dateFormatter.dateFormat = @"MM-dd HH:mm:ss ";
    
    NSString *log = [NSString stringWithFormat:@"%@ %@", [dateFormatter stringFromDate:now], message];
    
    dispatch_async(dispatch_get_main_queue(), ^{
        self.adLogView.text = [NSString stringWithFormat:@"%@\n%@", log, self.adLogView.text];
    });
    
    NSLog(@"[VAMP]%@", message);
}

- (NSString *)vampStateString:(VAMPState)state {
    switch (state) {
        case kVAMPStateIdle:
            return @"Idle";
        case kVAMPStateLoading:
            return @"Loading";
        case kVAMPStateReady:
            return @"Ready";
        case kVAMPStateLoaded:
            return @"Loaded";
        case kVAMPStateShowing:
            return @"Showing";
        default:
            return @"Unknown";
    }
}

#pragma mark - VAMPDelegate

// load完了して、広告表示できる状態になったことを通知します
- (void)vampDidReceive:(NSString *)placementId adnwName:(NSString *)adnwName {
    [self addLogText:[NSString stringWithFormat:@"vampDidReceive(%@) placementId:%@",
                      adnwName, placementId]];
}

// エラーが発生した時に通知されます
- (void)vampDidFail:(NSString *)placementId error:(VAMPError *)error {
    NSString *codeString = [error kVAMPErrorString];
    NSString *failMessage = error.localizedDescription;
    
    [self addLogText:[NSString stringWithFormat:@"vampDidFail placementId:%@ %@ %@",
                      placementId, codeString, failMessage]];
}

// インセンティブ付与可能になったタイミングで通知されます
- (void)vampDidComplete:(NSString *)placementId adnwName:(NSString *)adnwName {
    [self addLogText:[NSString stringWithFormat:@"vampDidComplete(%@) placementId:%@",
                      adnwName, placementId]];
}

// 広告が閉じられた時に呼ばれます
- (void)vampDidClose:(NSString *)placementId adnwName:(NSString *)adnwName {
    [self addLogText:[NSString stringWithFormat:@"vampDidClose(%@) placementId:%@",
                      adnwName, placementId]];
    
    if (self.isPlayingPrev) {
        [self.soundPlayer play];
    }
}

// アドネットワークごとの広告取得が開始された時に通知されます
- (void)vampLoadStart:(NSString *)placementId adnwName:(NSString *)adnwName {
    [self addLogText:[NSString stringWithFormat:@"vampLoadStart(%@) placementId:%@",
                      adnwName, placementId]];
}

// アドネットワークごとの広告取得結果が通知されます（success,failedどちらも通知）。
// この通知をもとにshowしないようご注意ください。showする判定は、vampDidReceiveを受け取ったタイミングで判断してください
- (void)vampLoadResult:(NSString *)placementId success:(BOOL)success adnwName:(NSString *)adnwName
               message:(NSString *)message {
    
    if (success) {
        [self addLogText:[NSString stringWithFormat:@"vampLoadResult(%@) success:%@ placementId:%@",
                          adnwName, success ? @"YES" : @"NO", placementId]];
    } else {
        [self addLogText:[NSString stringWithFormat:@"vampLoadResult(%@) success:%@ message:%@ placementId:%@",
                          adnwName, success ? @"YES" : @"NO", message, placementId]];
    }
}

// 広告準備完了から55分経つと取得した広告の表示はできてもRTBの収益は発生しません。
// この通知を受け取ったら、もう一度loadからやり直す必要があります
- (void)vampDidExpired:(NSString *)placementId {
    [self addLogText:[NSString stringWithFormat:@"vampDidExpired placementId:%@", placementId]];
}

// VAMPの状態が変化したときの通知されます
- (void)vampDidChangeState:(VAMPState)oldState intoState:(VAMPState)newState
          withPlacementId:(NSString *)placementId {
    
    NSString *oldStateStr = [self vampStateString:oldState];
    NSString *newStateStr = [self vampStateString:newState];
    
    [self addLogText:[NSString stringWithFormat:@"vampDidChangeState %@ -> %@, placementId:%@",
                      oldStateStr, newStateStr, placementId]];
}

@end

File: RewardSampleForSwift/RewardSampleForSwift/VideoSingleViewController.swift
//
//  VideoSingleViewController.swift
//  RewardSampleForSwift
//
//  Created by AdGeneration on 2017/02/21.
//  Copyright © 2017年 Supership Inc. All rights reserved.
//

import Foundation
import AVFoundation
import VAMP

class VideoSingleViewController: UIViewController, VAMPDelegate {
    
    @IBOutlet var adcodeField: UITextField!
    @IBOutlet var adLogView: UITextView!
    
    /**
     広告枠IDを設定してください
     59755 : iOSテスト用ID (このIDのままリリースしないでください)
     */
    let placementId = "59755"
    
    var vamp: VAMP!
    
    var soundOffButton: UIBarButtonItem!
    var soundOnButton: UIBarButtonItem!
    var soundPlayer: AVAudioPlayer!
    var isPlayingPrev = false
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        // テストモード確認するためログ出力
        print("[VAMP]isTestMode:" + String(VAMP.isTestMode()))

        // デバッグモード確認するためログ出力
        print("[VAMP]isDebugMode:" + String(VAMP.isDebugMode()))
        
        self.automaticallyAdjustsScrollViewInsets = false
        
        self.adcodeField.text = self.getPlacementId()
        
        self.adLogView.text = ""
        self.adLogView.isEditable = false
        
        let soundOnImage = UIImage(named: "soundOn")?.withRenderingMode(.alwaysOriginal)
        self.soundOffButton = UIBarButtonItem.init(image: soundOnImage, style: .plain, target: self,
                                                   action: #selector(self.soundOff))
        
        let soundOffImage = UIImage(named: "soundOff")?.withRenderingMode(.alwaysOriginal)
        self.soundOnButton = UIBarButtonItem.init(image: soundOffImage, style: .plain, target: self,
                                                  action: #selector(self.soundOn))
        
        self.navigationItem.rightBarButtonItem = self.soundOnButton
        
        let soundPath = Bundle.main.path(forResource: "invisible", ofType: "mp3")
        let soundUrl = URL(fileURLWithPath: soundPath!)
        
        do {
            // Fallback on earlier versions
            self.soundPlayer = try AVAudioPlayer(contentsOf: soundUrl)
            self.soundPlayer.prepareToPlay()
        } catch {
            print(error)
        }
        
        // VAMPインスタンスを生成し初期化します
        self.vamp = VAMP()
        self.vamp.delegate = self
        self.vamp.setPlacementId(self.getPlacementId())
        self.vamp.setRootViewController(self)
    }
    
    @objc func soundOff() {
        self.navigationItem.rightBarButtonItem = self.soundOnButton
        self.soundPlayer.pause()
    }
    
    @objc func soundOn() {
        self.navigationItem.rightBarButtonItem = self.soundOffButton
        self.soundPlayer.play()
    }
    
    func getPlacementId() -> String {
        return placementId
    }
    
    @IBAction func loadAd(sender: AnyObject) {
        self.addLogText("[load]")
        
        // 広告の読み込みを開始します
        self.vamp.load()
    }
    
    @IBAction func showAd(sender: AnyObject) {
        // 広告の準備ができているか確認してから表示してください
        if (self.vamp.isReady()) {
            self.addLogText("[show]")
            self.pauseSound()

            // 広告を表示
            self.vamp.show()
        } else {
            print("[VAMP]not ready")
        }
    }
    
    func addLogText(_ message: String, color: UIColor) {
        let dateFormatter = DateFormatter()
        dateFormatter.locale = NSLocale.system
        dateFormatter.dateFormat = "MM-dd HH:mm:ss"
        let timestamp = dateFormatter.string(from: Date())
        
        let attributedNow = NSAttributedString.init(string: String.init(format: "%@ ", timestamp), attributes: [NSAttributedStringKey.foregroundColor : UIColor.gray])
        let attributedMessage = NSAttributedString.init(string: String.init(format: "%@\n", message), attributes: [NSAttributedStringKey.foregroundColor : color])
        
        DispatchQueue.main.async() {
            let mutableAttributedString = NSMutableAttributedString.init()
            mutableAttributedString.append(attributedNow)
            mutableAttributedString.append(attributedMessage)
            mutableAttributedString.append(self.adLogView.attributedText)
            self.adLogView.attributedText = mutableAttributedString
        }
        
        print("[VAMP]\(timestamp) \(message)")
    }
    
    func addLogText(_ message: String) {
        self.addLogText(message, color: UIColor.gray)
    }
    
    func vampStateString(_ state: VAMPState) -> String {
        switch state {
        case .idle:
            return "Idle"
        case .loading:
            return "Loading"
        case .loaded:
            return "Loaded"
        case .showing:
            return "Showing"
        }
    }
    
    func pauseSound() {
        self.isPlayingPrev = self.soundPlayer.isPlaying
        
        if (self.soundPlayer.isPlaying) {
            self.soundPlayer.pause()
        }
    }
    
    func resumeSound() {
        if self.isPlayingPrev {
            self.soundPlayer.play()
        }
    }
    
    // MARK: - VAMPDelegate
    
    // 広告取得完了
    // 広告表示が可能になると通知されます
    func vampDidReceive(_ placementId: String, adnwName: String) {
        self.addLogText("vampDidReceive(\(adnwName), \(placementId))")
    }
    
    // 広告取得失敗
    // 広告が取得できなかったときに通知されます。
    // 例）在庫が無い、タイムアウトなど
    // @see https://github.com/AdGeneration/VAMP-iOS-SDK/wiki/VAMP-iOS-API-Errors
    func vamp(_ vamp: VAMP, didFailToLoadWithError error: VAMPError, withPlacementId placementId: String?) {
        if let bindPid = placementId {
            self.addLogText("vampDidFailToLoad(\(error.localizedDescription), \(bindPid))", color: UIColor.red)

            let code = VAMPErrorCode(rawValue: UInt(error.code))
            if code == .noAdStock {
                // 在庫が無いので、再度loadをしてもらう必要があります。
                // 連続で発生する場合、時間を置いてからloadをする必要があります。
                print("[VAMP]vampDidFailToLoad(noAdStock, \(error.localizedDescription))")
            } else if code == .noAdnetwork {
                // アドジェネ管理画面でアドネットワークの配信がONになっていない、
                // またはEU圏からのアクセスの場合(GDPR)に発生します。
                print("[VAMP]vampDidFailToLoad(noAdnetwork, \(error.localizedDescription))")
            } else if code == .needConnection {
                // ネットワークに接続できない状況です。
                // 電波状況をご確認ください。
                print("[VAMP]vampDidFailToLoad(needConnection, \(error.localizedDescription))")
            } else if code == .mediationTimeout {
                // アドネットワークSDKから返答が得られず、タイムアウトしました。
                print("[VAMP]vampDidFailToLoad(mediationTimeout, \(error.localizedDescription))")
            }
        }
    }
    
    // 広告表示失敗
    // show実行したが、何らかの理由で広告表示が失敗したときに通知されます。
    // 例）ユーザーが広告再生を途中でキャンセルなど
    func vamp(_ vamp: VAMP, didFailToShowWithError error: VAMPError, withPlacementId placementId: String) {
        self.addLogText("vampDidFailToShow(\(error.localizedDescription), \(placementId))", color: UIColor.red)
        
        let code = VAMPErrorCode(rawValue: UInt(error.code))
        if (code == .userCancel) {
            // ユーザが広告再生を途中でキャンセルしました。
            // AdMobは動画再生の途中でユーザーによるキャンセルが可能
            print("[VAMP]vampDidFailToShow(userCancel, \(error.localizedDescription))");
        } else if (code == .notLoadedAd) {
            print("[VAMP]vampDidFailToShow(notLoadedAd, \(error.localizedDescription))");
        }
        
        self.resumeSound()
    }
    
    // インセンティブ付与OK
    // インセンティブ付与が可能になったタイミングで通知
    // アドネットワークによって通知タイミングが異なる（動画再生完了時、またはエンドカードを閉じたタイミング）
    func vampDidComplete(_ placementId: String, adnwName: String) {
        self.addLogText("vampDidComplete(\(adnwName), \(placementId))", color: UIColor.blue)
    }
    
    // 広告閉じる
    // エンドカード閉じる、途中で広告再生キャンセル
    func vampDidClose(_ placementId: String, adnwName: String) {
        self.addLogText("vampDidClose(\(adnwName), \(placementId))", color: UIColor.black)
        self.resumeSound()
    }

    // 広告準備完了から55分経つと取得した広告の表示はできてもRTBの収益は発生しません。
    // この通知を受け取ったら、もう一度loadからやり直す必要があります
    func vampDidExpired(_ placementId: String) {
        self.addLogText("vampDidExpired(\(placementId))", color: UIColor.red)
    }

    // 広告取得開始
    // アドネットワークの広告取得が開始されたときに通知されます
    func vampLoadStart(_ placementId: String, adnwName: String) {
        self.addLogText("vampLoadStart(\(adnwName), \(placementId))")
    }
    
    // アドネットワークの広告取得結果が通知されます。成功時はsuccess=YESとなりロード処理は終了します。
    // success=NOのとき、次位のアドネットワークがある場合はロード処理は継続されます
    func vampLoadResult(_ placementId: String, success: Bool, adnwName: String, message: String?) {
        if success {
            self.addLogText("vampLoadResult(\(adnwName), \(placementId), success:OK)", color: UIColor.black)
        } else {
            let msg = message != nil ? message! : ""
            self.addLogText("vampLoadResult(\(adnwName), \(placementId), success:NG, \(msg))", color: UIColor.red)
        }
    }
    
    // 状態が変化したときに通知されます
    func vampDidChange(_ oldState: VAMPState, into newState: VAMPState, withPlacementId placementId: String?) {
//        let oldStateStr = self.vampStateString(oldState)
//        let newStateStr = self.vampStateString(newState)
//
//        self.addLogText("vampDidChangeState(\(oldStateStr) -> \(newStateStr), \(placementId))")
    }
}
